File: server/app/modules/articles/schema.py
from datetime import datetime

from pydantic import BaseModel, Field, HttpUrl


class ArticleSource(BaseModel):
    """Original publisher of the article."""

    name: str
    url: HttpUrl | None = None


class ArticleProvider(BaseModel):
    """API/provider that supplied the article."""

    name: str
    article_id: str | None = None


class FetchState(BaseModel):
    """Tracks when a category or tag was last fetched."""

    key: str
    last_fetched_at: datetime | None = None


class ArticleCreate(BaseModel):
    title: str
    description: str | None = None

    # Original article URL.
    # Used as the primary cross-provider deduplication key.
    url: HttpUrl | None = None

    image_url: HttpUrl | None = None

    source: ArticleSource

    # APIs that supplied this article.
    providers: list[ArticleProvider] = Field(default_factory=list)

    categories: list[str] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)

    published_at: datetime | None = None


class ArticleResponse(ArticleCreate):
    id: str
    created_at: datetime


class ArticleListResponse(BaseModel):
    articles: list[ArticleResponse]
    total: int
    page: int
    limit: int


class ArticleQuery(BaseModel):
    categories: list[str] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    limit: int = Field(default=10, ge=1, le=100)
    page: int = Field(default=1, ge=1)

File: client/src/features/articles/schema.ts
import { z } from "zod";

export const ArticleSourceSchema = z.object({
  name: z.string(),
  url: z.string().url().nullable().optional(),
});

export const ArticleProviderSchema = z.object({
  name: z.string(),
  article_id: z.string().nullable().optional(),
});

export const FetchStateSchema = z.object({
  key: z.string(),
  last_fetched_at: z.string().datetime().nullable().optional(),
});

export const ArticleCreateSchema = z.object({
  title: z.string(),
  description: z.string().nullable().optional(),

  url: z.string().url().nullable().optional(),

  image_url: z.string().url().nullable().optional(),

  source: ArticleSourceSchema,

  providers: z.array(ArticleProviderSchema).default([]),

  categories: z.array(z.string()).default([]),

  tags: z.array(z.string()).default([]),

  published_at: z.string().datetime().nullable().optional(),
});

export const ArticleResponseSchema = ArticleCreateSchema.extend({
  id: z.string(),
  created_at: z.string().datetime(),
});

export const ArticleListResponseSchema = z.object({
  articles: z.array(ArticleResponseSchema),
  total: z.number(),
  page: z.number(),
  limit: z.number(),
});

export const ArticleQuerySchema = z.object({
  categories: z.array(z.string()).default([]),
  tags: z.array(z.string()).default([]),
  limit: z.number().int().min(1).max(100).default(10),
  page: z.number().int().min(1).default(1),
});

export type ArticleSource = z.infer<typeof ArticleSourceSchema>;
export type ArticleProvider = z.infer<typeof ArticleProviderSchema>;
export type FetchState = z.infer<typeof FetchStateSchema>;
export type ArticleCreate = z.infer<typeof ArticleCreateSchema>;
export type ArticleResponse = z.infer<typeof ArticleResponseSchema>;
export type ArticleListResponse = z.infer<typeof ArticleListResponseSchema>;
export type ArticleQuery = z.infer<typeof ArticleQuerySchema>;
